File: src/main/java/com/nobility/downloader/settings/Defaults.java
package com.nobility.downloader.settings;

/**
 * An enum used for easy settings management.
 * The type is used to identify the type for the setting. This makes loading them faster.
 * The value is an object that is identified by the type.
 * 0 is for booleans
 * 1 is for integers
 * 2 is for strings
 */
public enum Defaults {

    DENIEDUPDATE("deniedupdate", 0, false),
    UPDATEVERSION("version", 2, "1.0"),
    PROXY("proxy", 2, ""),
    TIMEOUT("timeout", 1, 30),
    ENABLEPROXY("enable_proxy", 0, false),
    SAVELINKS("save_links", 0, false),
    SHOWCONTEXTONCLICK("show_context", 0, true),
    MAXEPISODES("max_episodes", 1, 30),
    NOEPISODELIMIT("no_episode_limit", 0, false),
    SILENTDRIVER("silent_driver", 0, true),
    SAVEFOLDER("save_folder", 2, System.getProperty("user.home")),
    DOWNLOADTHREADS("download_threads", 1, 2),
    LASTDOWNLOAD("last_dl", 2, ""),
    DRIVER("driver", 2, DriverDefaults.CHROME.name());

    private final String key;
    private final int type; //0 == boolean, 1 == int, 2 == string
    private final Object value;

    Defaults(String key, int type, Object value) {
        this.key = key;
        this.type = type;
        this.value = value;
    }

    public String getKey() {
        return key;
    }

    public int getType() {
        return type;
    }

    public Object getValue() {
        return value;
    }

}

File: src/main/java/com/nobility/downloader/settings/Defaults.kt
package com.nobility.downloader.settings

import com.nobility.downloader.driver.DriverDefaults

/**
 * An enum used for easy settings management.
 * The type is no longer needed with the new system.
 * Settings only support numbers, strings and booleans.
 * Custom saves must be implemented inside SettingsMeta
 * @see com.nobility.downloader.entities.settings.SettingsMeta
 */
enum class Defaults(
    val key: String,
    val value: Any
) {
    BYPASSFREESPACECHECK("bypass_fs_check", false),
    DENIEDUPDATE("deniedupdate", false),
    UPDATEVERSION("version", "1.0"),
    PROXY("proxy", ""),
    TIMEOUT("timeout", 30),
    ENABLEPROXY("enable_proxy", false),
    SHOWCONTEXTONCLICK("show_context", true),
    SILENTDRIVER("silent_driver", true),
    SAVEFOLDER("save_folder", System.getProperty("user.home")),
    DOWNLOADTHREADS("download_threads", 2),
    LASTDOWNLOAD("last_dl", ""),
    TOASTTRANSPARENCY("toast_trans", 50.toDouble()),
    DRIVER("driver", DriverDefaults.CHROME.name),
    DEBUGMESSAGES("debug_messages", false),
    DOMAIN("wco_website", "wcofun"), //used to combat the constantly changing domain
    EXTENSION("wco_ext", "org"), //used to combat the constantly changing domain
    QUALITY("video_quality", Quality.LOW.tag),
    NEW_DB_WARNING("new_db_warn", false)

}